### PR TITLE
feat: moves to building with lualatex

### PR DESCRIPTION
**Pull request by matsavage**
Moves from XeLaTeX to LuaLaTeX

Ideally running `latexmk` would work fine inside `flake.nix` without having to override the return code, but I can't seem to figure this out at the moment.

Will return 0 either way, but the `installPhase` will fail if no PDF was generated, so feels like a small problem

### CODE REVIEW

**Comment by PJBrs** [comment]
For transparancies, the first lines of the document need to read:
```
\RequirePackage{pdfmanagement-testphase}
\DocumentMetadata{}
```

**Comment by PJBrs** [comment]
Hi, so I've given the whole "which latex engine to use" question some more thought, especially with an eye to dungeonsheets integration. Here's a list of cons and pros for various options that I tried.

**xelatex**
Known to be good, works with fontspec, so can use kalam font as is. However, it is very slow and needs a way to pass NOSAFER to xdvipdfmx. I've managed to get this to run with dungeonsheets, with the following patch (it's a kludge):
```
--- a/dungeonsheets/latex.py
+++ b/dungeonsheets/latex.py
@@ -69,6 +69,9 @@ def create_latex_pdf(
         str(tex_file),
     ]
 
+    if "xelatex" in tex_command_line:
+        tex_command_line.insert(1, '--no-pdf')
+
     environment = os.environ
     tex_env = environment.get('TEXINPUTS', '')
     module_root = Path(__file__).parent / "modules/"
@@ -90,6 +93,12 @@ def create_latex_pdf(
             result = subprocess.run(
                 tex_command_line, stdout=subprocess.DEVNULL, env=environment, timeout=30
             )
+        if "xelatex" in tex_command_line:
+            result = subprocess.run(
+                ["xdvipdfmx", "-i", "dvipdfmx-unsafe.cfg", "-q", "-E",
+                str(tex_file).replace('.tex', '.xdv'),],
+                stdout=subprocess.DEVNULL,  env=environment, timeout=30
+            )
     except FileNotFoundError:
         # Remove temporary files
         _remove_temp_files(basename)
diff --git a/dungeonsheets/make_sheets.py b/dungeonsheets/make_sheets.py
index 7757613..fcab23d 100644
--- a/dungeonsheets/make_sheets.py
+++ b/dungeonsheets/make_sheets.py
@@ -546,7 +546,7 @@ def msavage_sheet(character, basename, debug=False):
         basename=basename,
         keep_temp_files=debug,
         use_dnd_decorations=True,
-        comm1="lualatex",
+        comm1="xelatex",
```
You'd think it ought to have been possible to use xelatex by passing `-output-driver="xdvipdfmx -i dvipdfmx-unsafe.cfg -q -E"` as part of `tex_command_line` and compile in one go, but it fails with "xdvipdfmx -i dvipdfmx-unsafe.cfg -q -E" not found or something.

**lualatex, with luapstricks**
Lualatex is great and fast, and compatible with fontspec, so no Kalam font issues! I managed to get it to run on Kubuntu Jammy as well, but it adds a blank first page. This (probably) means that only rather new texlive distributions will work...

**lualatex or pdflatex using auto-pst-pdf**
 So the whole issue of xelatex going slow goes back to pstricks. And the whole issue of lualatex requiring very new versions as well. So we might instead use the much older auto-pst-pdf as will, which is compatible with both pdflatex and lualatex (and probably xelatex, but I didn't try that). It requires the following in the characterfile:

```
\usepackage{ifpdf}
\ifpdf
\usepackage[pspdf=-dALLOWPTRANSPARENCY]{auto-pst-pdf}
\else
\usepackage{pstricks}
\fi
```
 The problem here is that, in this case, I can't combine pstricks with fontspec, and since we use \entryfont with \rput, we mix pstricks with fontspec, and the character fails to compile. I managed to compile this by commenting out `\usepackage{fontspect}`  and the newfontfamily line in dndtemplate.sty, and adding `\newcommand\entryfont`. This compiles very fast, can even use pdflatex  (!), but doesn't have the kalam font. If, however, we could convert the kalam font to a type that is compatible with pdflatex (and set the TFMFONT environment variable to find out), this might actually be the fastest, most robust, and most backward-compatible option.

So, I'm going to try and convert kalam.ttf to something pdflatex can grasp.

To sum up: my current dungeonsheets  branch with lualatex works like it should, by only on relatively new distros. Alternatively, dungeonsheets can be made to successfully use xelatex as well, but it requires either external configuration or the above kludge. The auto-pst-pdf route looks really promising but requires some font work.

**Comment by PJBrs** [comment]
Just a note to self.
`TFMFONTS=template/fonts:$TFMFONTS ENCFONTS=template/fonts:$ENCFONTS T1FONTS=template/fonts:$T1FONTS
`

**Comment by matsavage** [comment]
I might be able to put some time into this tomorrow, but have been thinking today. If the issue with lualatex is the transparency, might it not just be simpler to have an if statement for the white backgrounds, then it’ll just be there or not. No transparency needed. @PJBrs any thoughts?

**Comment by PJBrs** [comment]
I've been doing some more testing with other compilers. I think that I have exhausted all options with 1) pdflatex and auto-pst-pdf  and 2) latex->dvips->ps2pdf. Both have big problems, such as only being able to deal with jpg images, and it's very hard to use ttf fonts combined with dvips.

I've had one more thought though... It seems to me that only a limited number of boxes in the character sheet actually have a transparency, the others appear to consist of lines only. How hard would it be to find the boxes that are filled with transparency? In that case, compiling would be all the more simple.

I've also been thinking about overflowing text, and I stumbled on at least three ways to better constrain text, both horizontally and vertically:

https://tex.stackexchange.com/questions/374032/how-to-resize-font-into-box-automatically

https://tex.stackexchange.com/questions/123614/making-the-text-fit-in-a-specific-space-in-latex

**Comment by PJBrs** [comment]
Re. transparencies...  Some first testing suggests that, instead of setting opacity (which requires allowing transparency, it might be easier to distinguish between:

\pscustom[linestyle=none,fillstyle=solid,fillcolor=curcolor]

and

\pscustom[linestyle=none,fillstyle=none

instead of setting opacity itself. You'd keep the same amount of customisation, but without the hassle of requiring transparency.

**Comment by PJBrs** [comment]
Re. transparency - sth like this will work:
```
\ifnum \generalbackgroundopacity=0
  \def \fill{none}
\else
  \def \fill{solid}
\fi

{
\newrgbcolor{curcolor}{1 1 1}
\pscustom[linestyle=none,fillstyle=\fill,fillcolor=curcolor]

```

This way, you avoid setting transparency anywhere, while keeping your customisability.

I'll prepare a patch for this.

**Comment by matsavage** [comment]
Yes this is sort of what I was getting at when I was suggesting that transparency could be a true/false if statement, but I hadn’t really thought about what the implementation might look like.

I’m very excited to have a look at some
of these text constraints. Is there a way to have the whole box rendered somewhere else, eg in a paragraph section on a following page and an overview on the page. 

I’m really not the LaTeX expert you are!

**Comment by PJBrs** [comment]
Definitely _not_ a latex expert, mind you! 

In any case, the latest test failed, but the results actually are very similar:
```
./result/background-sheet.pdf: difference from reference: 0.0265% 
./result/character-sheet.pdf: difference from reference: 0.0527% 
./result/full-caster-spell-sheet.pdf: difference from reference: 0.4184% 
./result/half-caster-spell-sheet.pdf: difference from reference: 0.4500% 
```

I downloaded them, and they actually look fine to me:

https://github.com/matsavage/DND-5e-LaTeX-Character-Sheet-Template/actions/runs/8316190594/artifacts/1333104505

Perhaps the threshold is too tight... I'm of a mind to update the tests ;-)

**Comment by PJBrs** [comment]
... and if I run locally, I get this :-)
$ ./scripts/check_test_output.sh 

./result/background-sheet.pdf: difference from reference: 0.0265% 
./result/character-sheet.pdf: difference from reference: 0.0527% 
./result/full-caster-spell-sheet.pdf: difference from reference: 0.4184% 
./result/half-caster-spell-sheet.pdf: difference from reference: 0.4500%

**Comment by PJBrs** [comment]
... and when I download the results and compare with my local run:
./result/background-sheet.pdf: difference from reference: 0.0000% 
./result/character-sheet.pdf: difference from reference: 0.0000% 
./result/full-caster-spell-sheet.pdf: difference from reference: 0.0000% 
./result/half-caster-spell-sheet.pdf: difference from reference: 0.0000%

**Comment by PJBrs** [comment]
OK, so no matter how I compile - xelatex, xelatex --output-driver="xdvipdfmx -i dvipdfmx-unsafe.cfg -q -E", or lualuatex with or without setting \DocumentMetadata{} - I get the same test result. I'm updating the reference :-)

**Comment by PJBrs** [comment]
Just to be clear, I'm talking about the test over here: https://github.com/matsavage/DND-5e-LaTeX-Character-Sheet-Template/pull/65

**Comment by PJBrs** [comment]
Re. Transparency -  checked out tcolorbox, and I think it can very well do what we need it to do. However, all the locations need to be moved down by a couple of points, the heights need to be found manually, and it gives an option clash with the dnd class, which already loads tcolorbox. So you need sth. like:

\PassOptionsToPackage{fitting}{tcolorbox}

and this needs to be passed before dnd is loaded, which in turn means that:

1. The character file needs to add \PassOptionsToPackage{fitting}{tcolorbox} before \usepackage{dnd}
2. The dndtemplate.sty needs to explicitly \PassOptionsToPackage{fitting}{tcolorbox} _and_ \RequirePackage{dnd} (which would result in an option clash if the user wants to pass options to dnd themselves.
3. The dndtemplate.sty needs to explicitly \PassOptionsToPackage{fitting}{tcolorbox} _and_ the character file needs to  \usepackage{dndtemplate} _before \usepackage{dnd}

The third option would be the best, _and_ it would require additional patching in dungeonsheets.

I'll also investigate the solution here:
https://tex.stackexchange.com/questions/26002/fit-text-into-given-box-by-adjusting-the-fontsize

**Comment by PJBrs** [comment]
I tested both options and they seem to work equally well, see over here:

https://github.com/PJBrs/DND-5e-LaTeX-Character-Sheet-Template/tree/fontscale

Especially the unnamed.tex file can be used for testing, just add more paragraphs of `\lipsum`.

For the character sheet, I've used a "fitbox" environment, the code for which I took from [stackexchange](https://tex.stackexchange.com/questions/26002/fit-text-into-given-box-by-adjusting-the-fontsize).

For the background sheet, I've used tcolorbox. 

@matsavage I'd like to prepare a patch for the entire document, if you'd like to have one. I personally like the fitbox option better. And on the character sheet I would add it for name, class, background and speed as well, so that long background names do not overflow player names, so that multiple classes can all be mentioned without overflowing the background, same holds for race and alignment, and speed could read sth. like 30 (swim 30).

**Comment by matsavage** [comment]
Hi PJ

That sounds great, I would be very happy to accept something like that, for my understanding using the fitbox allows for this to be done without any changes requires for the DnD template?

**Comment by PJBrs** [comment]
To be precise - the problem with the tcolorbox solution is that it generates an option clash with dnd.sty, because we would need to load tcolorbox with the "fitting" option, whereas dnd.sty loads it without that option. The most correct way to solve that (I've read some more about option clashes), is that the character file specifies `\PassOptionsToPackage{fitting}{tcolorbox}` _before_ it loads dnd.sty. This requires an additional change in the dungeonsheets template. After that, everything works as it should.

The advantage of the fitbox solution is that it doesn't draw in any new code, so it doesn't create additional dependencies (apart from `environ`, but that one's would have been pulled in by tcolorbox as well, and therefore also is loaded by dnd.sty.

The only thing that might be worrisome is that the fitbox code is not maintained anywhere, so that might pose a risk for the future, whereas, I assume, tcolorbox still receives active development.

**Comment by matsavage** [comment]
I think we can go with the easier option of including the external code here, and we can revisit if something eventually breaks

**Comment by matsavage** [comment]
I think this was all covered off by @PJBrs
